src/ruby/puppet-server-lib/puppet/server/master.rb

```diff
@@ -57,6 +60,53 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def getClassInfoForAllEnvironments()
```

Comment:
One other thing I had forgotten, though...

When getting `Environment` instances back from the `@env_loader.list`, the cache expiration logic is not implicitly executed.  That only happens when `@env_loader.get(env)` is called.  So I think the `Puppet.settings.clear_environment_settings()` calls are safer to make regardless of whether the environments in the cache had been marked for expiration just before this code is invoked.